File: kotlin-jetbrains-tutorial/src/classes/generics/java/Source.java
package classes.generics.java;

public class Source<T> {

    private T value;

    public Source(T value) {
        this.value = value;
    }

    public T nextValue() {
        return value;
    }

    public void setValue(T value) {
        this.value = value;
    }

//    public static void main(String[] args) {
//        show(new Source<>(1.12131));
//    }
//
//    public static <T extends Number> void show(Source<T> x) {
//        Source<Double> y = x; // This is completely safe since Double is a subtype of Number
//    }

}

File: kotlin-jetbrains-tutorial/src/classes/generics/classes/Source.kt
package classes.generics.classes

//interface Source<out T> {
//
//    fun nextValue(): T
//
//}

class Source<out T>(val value: T) {

    fun nextValue() = value

}

fun demo(source: Source<String>) {
    val newSource: Source<Any> = source

}